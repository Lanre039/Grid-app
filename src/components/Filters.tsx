import React, { useState } from "react";
import items from "../data/items.json";
import { FilterOption, FilterProps } from "../pages/types";

function Filters(props: FilterProps) {
  const [activeColor, setActiveColor] = useState<string[]>([]);
  const [activeShape, setActiveShape] = useState<string[]>([]);
  const { filter, handleFilters } = props;

  // Color would serve as the filter control
  // Remove color dublicate here
  const colorFilterItem = Array.from(new Set(items.map(({ color }) => color)));

  const handleClick = (filter: string, { shape = "", color = "" }) => {
    if (color) {
      handleFilters({
        color,
        category: filter,
      } as FilterOption);

      // SET ACTIVE FILTER - COLOR
      if (activeColor.includes(color)) {
        const activeItems = activeColor.filter((item) => item !== color);
        setActiveColor(activeItems);
      } else {
        setActiveColor([...activeColor, color]);
      }
    }

    if (shape) {
      handleFilters({
        shape,
        category: filter,
      } as FilterOption);

      // SET ACTIVE FILTER - SHAPE
      if (activeShape.includes(shape)) {
        const activeItems = activeShape.filter((item) => item !== shape);
        setActiveShape(activeItems);
      } else {
        setActiveShape([...activeShape, shape]);
      }
    }
  };

  const renderByShape = () => {
    const shapeFilterItem = items.filter(
      ({ color }) => color === colorFilterItem[0]
    );

    return shapeFilterItem.map(({ shape }, index) => {
      const active = activeShape.includes(shape)
        ? "filter_shape-active"
        : "filter_shape-inactive";

      return (
        <span
          key={index}
          className={`filter_container-item ${active}`}
          onClick={() => handleClick(filter, { shape })}
        >
          {shape}
        </span>
      );
    });
  };
  const renderByColor = () =>
    colorFilterItem.map((color, index) => {
      const active = activeColor.includes(color)
        ? "filter_color-active"
        : "filter_color-inactive";

      return (
        <span
          key={index}
          className={`filter_container-shapes ${color} ${active}`}
          onClick={() => handleClick(filter, { color })}
        ></span>
      );
    });

  return (
    <div className="filter_container">
      {filter === "shapes" ? renderByShape() : renderByColor()}
    </div>
  );
}

export default Filters;
