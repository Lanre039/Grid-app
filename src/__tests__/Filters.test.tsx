import { cleanup } from "@testing-library/react";
import { shallow, mount } from "enzyme";
import Filters from "../components/Filters";

afterEach(cleanup);

const props = {
  filter: "shapes",
  handleFilters: jest.fn(),
};

describe("<Filter />", () => {
  it("render correctly", () => {
    const wrapper = shallow(<Filters {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe("filter behaviour with shapes", () => {
    const noOfShapes = 5;
    const wrapper = mount(<Filters {...props} />);
    it("render number of span (shapes) to be 5", () => {
      const elements = wrapper.find("span");
      expect(elements.length).toBe(noOfShapes);
      expect(elements.at(0).text()).toBeTruthy();
    });

    it("span should have a text", () => {
      const elements = wrapper.find("span");
      expect(elements.at(0).text()).toBeTruthy();
    });

    it("add active to rounded shape after onClick event", () => {
      const shape = wrapper.find("span").first();

      expect(shape.html()).toBe(
        '<span class="filter_container-item filter_shape-inactive">rounded</span>'
      );

      shape.simulate("click");

      expect(props.handleFilters).toHaveBeenCalledTimes(1);

      expect(shape.html()).toBe(
        '<span class="filter_container-item filter_shape-active">rounded</span>'
      );
    });
  });

  describe("filter behaviour with color", () => {
    const noOfColors = 6;
    const wrapper = shallow(<Filters {...props} filter="color" />);

    it("render number of colors to be 6", () => {
      const elements = wrapper.find("span");

      expect(elements).toHaveLength(noOfColors);
    });

    it("span should NOT have a text", () => {
      const elements = wrapper.find("span");
      expect(elements.at(0).text()).toBeFalsy();
    });
  });
});
