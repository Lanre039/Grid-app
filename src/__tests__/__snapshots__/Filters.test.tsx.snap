// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> render correctly 1`] = `
<div
  className="filter_container"
>
  <span
    className="filter_container-item filter_shape-inactive"
    key="0"
    onClick={[Function]}
  >
    rounded
  </span>
  <span
    className="filter_container-item filter_shape-inactive"
    key="1"
    onClick={[Function]}
  >
    oval
  </span>
  <span
    className="filter_container-item filter_shape-inactive"
    key="2"
    onClick={[Function]}
  >
    triangle
  </span>
  <span
    className="filter_container-item filter_shape-inactive"
    key="3"
    onClick={[Function]}
  >
    square
  </span>
  <span
    className="filter_container-item filter_shape-inactive"
    key="4"
    onClick={[Function]}
  >
    rectangle
  </span>
</div>
`;
